packages/block-editor/src/components/link-control/input-search.js

```diff
@@ -0,0 +1,69 @@
+
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import {
+	IconButton,
+} from '@wordpress/components';
+import {
+	LEFT,
+	RIGHT,
+	UP,
+	DOWN,
+	BACKSPACE,
+	ENTER,
+} from '@wordpress/keycodes';
+
+/**
+ * Internal dependencies
+ */
+import {
+	URLInput,
+} from '../';
+
+const LinkControlInputSearch = ( { value, onChange, onSelect, renderSuggestions, fetchSuggestions, onReset } ) => {
+	const stopPropagation = ( event ) => {
+		event.stopPropagation();
+	};
+
+	const stopPropagationRelevantKeys = ( event ) => {
+		if ( [ LEFT, DOWN, RIGHT, UP, BACKSPACE, ENTER ].indexOf( event.keyCode ) > -1 ) {
```

Comment:
> ... stopPropagation needs to be handled when URLInput already handles it.

Not really. The `<URLInput /> doesn't block the propagation of the events and depending on the context when it's used the popover can disappear (apparently) spontaneously. You can see how it's handled in https://github.com/WordPress/gutenberg/pull/17986/ PR, [here](https://github.com/WordPress/gutenberg/pull/17986/files#diff-17e2993cb0b8cf64434015038b235162R59-R64) for instance.

There are some specific events that need to be handled in the `<URLInput />` component (like it already does). It's nice to see how the caret goes at the beginning or at the end when the user press `UP` or `DOWN`.

But handling the `LEFT` and `RIGHT` keys maybe need to be handled outside of the `<URLInput />`, since this is an special case where the component is used into a NavigationMenuItem, and we need to block the shifting to the next ones.

> At the very least these props need to be all marked as experimental.

I think adding the `__experimental` prefix to `onKeyPress` or `onKeyDown` could add more confusion. Why not simply passing the props and make them optional? 

```es6
<URLInput
  onKeyDown={ onKeyDownHandler }
  onKeyPress={ onKeyPressHandler }
/>
```

Reply:
...but I suggest handling it by the component which consumes it. For instance, in the NavigationMenuItem:

```es6
const NavigationMenuItem = ( props ) => {
	// ...
	return (
		<Toolbar>
			<ToolbarButton
				// ...
			/>
			<LinkControl
				className="wp-block-navigation-menu-item__inline-link-input"
				onKeyDown={ onKeyDownHandler }
				onKeyPress={ onKeyPressHandler }
			/>
		</Toolbar>
	);
};
```

Reply:
Let me show how we are handling this in the `[draft]` PR.

* [Removed the handling](https://github.com/WordPress/gutenberg/pull/18062/commits/59b4fdb542f400e0a02b85f3a1e2ce58ba25ae5a) code from the `<LinkControl />` component.
* Handle it straight [in the Item component](https://github.com/WordPress/gutenberg/pull/18062/commits/0a5c54f548de060dc91d0c1127774924564c338e). Just FYI, the bound onKeyDown function was written by @tellthemachines [on its PR](https://github.com/WordPress/gutenberg/pull/17986).

Reply:
@retrofox `URLInput` has been in use for a long time in popovers without those issues. The component does stop propagation of the events that are relevant to it:
https://github.com/WordPress/gutenberg/blob/6cf43ac11464cb182f64be7a9b4263a7e0f832c3/packages/block-editor/src/components/url-input/index.js#L175-L208

Other events do seem to be stopped at the form/popover level where `URLPopover.LinkEditor` is implemented. I do agree that there might be better ways to do this as it's a lot of code duplication. The issue would probably be that URLInput can be used both in and outside of a popover, so it shouldn't be handling this functionality internally.

> I think adding the __experimental prefix to onKeyPress or onKeyDown could add more confusion. Why not simply passing the props and make them optional?

I'm not sure I see why it would add more confusion. It signals that they're not (yet) for use externally, which is the truth seeing as we don't seem to have agreement on how this should be implemented. At the moment it would be a pragmatic way to get this PR shipped.

Reply:
> @retrofox URLInput has been in use for a long time in popovers without those issues.

So far 😆. It happens when we introduce new features. Simply happens. 

> Other events do seem to be stopped at the form/popover level where URLPopover.LinkEditor is implemented

Yes, it's true. I'm just wondering why not just pass the `onKeyPress` to the `<URLInput />` component, making it optional defining a `noop` function as default.

> I'm not sure I see why it would add more confusion. 

To me, it is the opposite. Simplifies the API instead of adding a new wrapper element just for subscribing to the event. So the options are:

1) Add a wrapper element to `<URLInput />` just to catch the event propagation.
2) Allow binding a function through of the `onKeyPress` property of the `<URLInput >`

> ...seeing as we don't seem to have agreement on how this should be implemented.

It's just part of the workflow, right? Totally agree to go for whatever we want and go ahead not blocking us. Personally, I prefer the second one but it's up to us.[ I've already added another comment about this](https://github.com/WordPress/gutenberg/pull/17846#issuecomment-545899012).